chain/chain/src/flat_storage_resharder.rs

```diff
@@ -0,0 +1,787 @@
+    let flat_store = resharder.runtime.store().flat_store();
+    let mut store_update = flat_store.store_update();
+    if success {
+        // Split shard completed successfully.
+        // Parent flat storage can be later deleted.
+        store_update.set_flat_storage_status(
+            parent_shard,
+            FlatStorageStatus::Resharding(FlatStorageReshardingStatus::ToBeDeleted),
+        );
+        // TODO(trisfald): trigger parent delete
+        // Children must perform catchup.
+        for shard in [left_child_shard, right_child_shard] {
+            store_update.set_flat_storage_status(
+                shard,
+                FlatStorageStatus::Resharding(FlatStorageReshardingStatus::CatchingUp(block_hash)),
+            );
+        }
+        // TODO(trisfald): trigger catchup
+    } else {
+        // We got an error or an interrupt request.
+        // Remove children shards leftovers and reset parent shard status.
+        // TODO(Trisfald): implement
+    }
+    store_update.commit().unwrap();
+    // Terminate the resharding event.
+    *resharder.resharding_event.lock().unwrap() = None;
+}
+
+/// Struct to describe, perform and track progress of a flat storage resharding.
+#[derive(Clone, Debug)]
+pub enum FlatStorageReshardingEvent {
+    /// Split a shard.
+    /// Includes the parent shard uid and the operation' status.
+    Split(ShardUId, SplittingParentStatus),
+}
+
+/// Helps control the flat storage resharder operation. More specifically,
+/// it has a way to know when the background task is done or to interrupt it.
+#[derive(Clone)]
+pub struct FlatStorageResharderController {
+    /// Resharding handle to control interruption.
+    handle: ReshardingHandle,
+    /// This object will be used to signal when the background task is completed.
+    /// A value of `true` means that the operation completed successfully.
+    completion_sender: Sender<bool>,
+    /// Corresponding receiver for `completion_sender`.
+    pub completion_receiver: Receiver<bool>,
+}
+
+impl FlatStorageResharderController {
+    /// Creates a new `FlatStorageResharderController` with its own handle.
+    pub fn new() -> Self {
+        let (completion_sender, completion_receiver) = crossbeam_channel::bounded(1);
+        let handle = ReshardingHandle::new();
+        Self { handle, completion_sender, completion_receiver }
+    }
+
+    pub fn from_resharding_handle(handle: ReshardingHandle) -> Self {
+        let (completion_sender, completion_receiver) = crossbeam_channel::bounded(1);
+        Self { handle, completion_sender, completion_receiver }
+    }
+
+    pub fn handle(&self) -> &ReshardingHandle {
+        &self.handle
+    }
+
+    /// Returns whether or not background task is interrupted.
+    pub fn is_interrupted(&self) -> bool {
+        !self.handle.get()
+    }
+}
+
+/// Represent the capability of scheduling the background tasks spawned by flat storage resharding.
+pub trait FlatStorageResharderScheduler {
```

Comment:
I came up with this trait to abstract the task scheduling. It works well in tests, I hope it can be plugged nicely with actors or whatever else we will end up using.